staging/src/k8s.io/component-base/featuregate/feature_gate.go

```diff
@@ -422,19 +422,50 @@ func (f *featureGate) AddVersioned(features map[Feature]VersionedSpecs) error {
 	if f.closed {
 		return fmt.Errorf("cannot add a feature gate after adding it to the flag set")
 	}
-
 	// Copy existing state
 	known := f.GetAllVersioned()
 
 	for name, specs := range features {
-		sort.Sort(specs)
 		if existingSpec, found := known[name]; found {
 			sort.Sort(existingSpec)
 			if reflect.DeepEqual(existingSpec, specs) {
 				continue
 			}
 			return fmt.Errorf("feature gate %q with different spec already exists: %v", name, existingSpec)
 		}
+		// Validate new specs are well-formed
+		var lastVersion *version.Version
+		var wasBeta, wasGA, wasDeprecated bool
+		for i, spec := range specs {
+			if spec.Version == nil {
+				return fmt.Errorf("feature %q did not provide a version", name)
+			}
+			// gates that begin as deprecated must indicate their prior state
+			if i == 0 && spec.PreRelease == Deprecated && spec.Version.Minor() != 0 {
+				return fmt.Errorf("feature %q introduced as deprecated must provide a 1.0 entry indicating initial state", name)
+			}
+			// Deprecated state must be the terminal state
+			if wasDeprecated && spec.PreRelease != Deprecated {
+				return fmt.Errorf("deprecated feature %q must not resurrect from its terminal state", name)
```

Comment:
nit: could move this into the switch below alongside other stability regression checks

```
case spec.PreRelease != Deprecated && wasDeprecated:
  return fmt.Errorf("deprecated feature %q must not resurrect from its terminal state", name)
```

Reply:
done